brian2genn/templates/engine.cpp

```diff
@@ -108,46 +108,106 @@ void engine::run(double duration, //!< Duration of time to run the model for
   double elapsed_realtime;
 
   for (int i= 0; i < riT; i++) {
-      // report state
-      {% for sm in state_monitor_models %}
-      {% if sm.when == 'start' %}
-      {% for var in sm.variables %}
-      {% if sm.isSynaptic %}
-      {% if sm.connectivity == 'DENSE' %}
-      convert_dense_matrix_2_dynamic_arrays({{var}}{{sm.monitored}}, {{sm.srcN}}, {{sm.trgN}},brian::_dynamic_array_{{sm.monitored}}__synaptic_pre, brian::_dynamic_array_{{sm.monitored}}__synaptic_post, brian::_dynamic_array_{{sm.monitored}}_{{var}});
+      // The StateMonitor and run_regularly operations are ordered by their "order" value
+      {% for is_state_monitor, obj in run_reg_state_monitor_operations %}
+      {% if is_state_monitor %}
+      {% if obj.when == 'start' %}
+      {% for var in obj.variables %}
+      {% if obj.isSynaptic %}
+      {% if obj.connectivity == 'DENSE' %}
+      convert_dense_matrix_2_dynamic_arrays({{var}}{{obj.monitored}},
+                                            {{obj.srcN}}, {{obj.trgN}},
+                                            brian::_dynamic_array_{{obj.monitored}}__synaptic_pre,
+                                            brian::_dynamic_array_{{obj.monitored}}__synaptic_post,
+                                            brian::_dynamic_array_{{obj.monitored}}_{{var}});
       {% else %}
-      convert_sparse_synapses_2_dynamic_arrays(C{{sm.monitored}}, {{var}}{{sm.monitored}}, {{sm.srcN}}, {{sm.trgN}}, brian::_dynamic_array_{{sm.monitored}}__synaptic_pre, brian::_dynamic_array_{{sm.monitored}}__synaptic_post, brian::_dynamic_array_{{sm.monitored}}_{{var}}, b2g::FULL_MONTY);
+      convert_sparse_synapses_2_dynamic_arrays(C{{obj.monitored}},
+                                               {{var}}{{obj.monitored}},
+                                               {{obj.srcN}}, {{obj.trgN}},
+                                               brian::_dynamic_array_{{obj.monitored}}__synaptic_pre,
+                                               brian::_dynamic_array_{{obj.monitored}}__synaptic_post,
+                                               brian::_dynamic_array_{{obj.monitored}}_{{var}},
+                                               b2g::FULL_MONTY);
       {% endif %}
       {% else %}
-      copy_genn_to_brian({{var}}{{sm.monitored}}, brian::_array_{{sm.monitored}}_{{var}}, {{sm.N}});
+      copy_genn_to_brian({{var}}{{obj.monitored}}, brian::_array_{{obj.monitored}}_{{var}}, {{obj.N}});
       {% endif %}
       {% endfor %}
-      _run_{{sm.name}}_codeobject();
+      _run_{{obj.name}}_codeobject();
       {% endif %}
-      {% endfor %}
-      // Execute scalar code for run_regularly operations (if any)
-      {% for nm in neuron_models %}
-      {% if nm.run_regularly_object != None %}
-      if (i % {{nm.run_regularly_step}} == 0)
+      {% else %}
+      if (i % {{obj['step']}} == 0)
```

Comment:
Just to be sure - is the intended behaviour that run_regulary timing is based on the local iteration variable i? Rather than on the global integer time step iT? I can't remember how we determine riT and whether the loop would ever be executed multiple times?

Reply:
We currently do not allow for multiple runs, so `i` and `iT` should always be identical. I think I prefer to use `i` here, since `iT` is a bit less straightforward (it gets updated as part of `stepTimeGPU`/`stepTimeCPU`, etc.).